game/gamestate.go

```diff
@@ -40,13 +55,154 @@ func (gs *GameState) Copy() *GameState {
 	}
 }
 
+// LegalMoves returns all legal moves for the current player.
+func (gs *GameState) LegalMoves() []Move {
+	switch gs.Phase {
+	case ReinforcementPhase:
+		return gs.reinforcementMoves()
+	case AttackPhase:
+		return gs.attackMoves()
+	case ManeuverPhase:
+		return gs.maneuverMoves()
+	default:
+		return nil
+	}
+}
+
+// reinforcementMoves generates all possible reinforcement moves for the current player.
+func (gs *GameState) reinforcementMoves() []Move {
+	var moves []Move
+	remainingTroops := gs.TroopsToPlace
+
+	// Get enemy-adjacent territories
+	enemyAdjacentTerritories := gs.getEnemyAdjacentTerritories()
+
+	// Possible troop amounts: one, half, all
+	troopAmounts := []int{1, remainingTroops / 2, remainingTroops}
+
+	for _, cantonID := range enemyAdjacentTerritories {
+		for _, amount := range troopAmounts {
+			if amount > 0 && amount <= remainingTroops {
+				moves = append(moves, &GameMove{
+					ActionType: ReinforceAction,
+					ToCantonID: cantonID,
+					NumTroops:  amount,
+				})
+			}
+		}
+	}
+	return moves
+}
+
+func (gs *GameState) attackMoves() []Move {
+	var moves []Move
+
+	for cantonID, owner := range gs.Ownership {
+		if owner == gs.CurrentPlayer && gs.TroopCounts[cantonID] > 1 {
+			maxTroops := min(gs.TroopCounts[cantonID]-1, gs.Rules.MaxAttackTroops())
+			for numTroops := 1; numTroops <= maxTroops; numTroops++ {
+				for _, adjID := range gs.Map.Cantons[cantonID].AdjacentIDs {
+					if gs.Ownership[adjID] != gs.CurrentPlayer {
+						moves = append(moves, &GameMove{
+							ActionType:   AttackAction,
+							FromCantonID: cantonID,
+							ToCantonID:   adjID,
+							NumTroops:    numTroops,
+						})
+					}
+				}
+			}
+		}
+	}
+
+	moves = append(moves, &GameMove{
+		ActionType: PassAction,
+	})
+	return moves
+}
+
+func (gs *GameState) getPlayerTerritoriesWithTroops() []int {
+	var territories []int
+	for cantonID, owner := range gs.Ownership {
+		if owner == gs.CurrentPlayer && gs.TroopCounts[cantonID] > 1 {
+			territories = append(territories, cantonID)
+		}
+	}
+	return territories
+}
+
+// maneuverMoves generates all possible maneuver moves for the current player.
+func (gs *GameState) getEnemyAdjacentOwnedTerritories() []int {
+	territoriesSet := make(map[int]struct{})
+	for cantonID, owner := range gs.Ownership {
+		if owner == gs.CurrentPlayer {
+			for _, adjID := range gs.Map.Cantons[cantonID].AdjacentIDs {
+				if gs.Ownership[adjID] != gs.CurrentPlayer {
+					territoriesSet[cantonID] = struct{}{}
+					break
+				}
+			}
+		}
+	}
+	var territories []int
+	for t := range territoriesSet {
+		territories = append(territories, t)
+	}
+	return territories
+}
+
+func (gs *GameState) maneuverMoves() []Move {
+	var moves []Move
+
+	fromTerritories := gs.getPlayerTerritoriesWithTroops()
+	toTerritories := gs.getEnemyAdjacentOwnedTerritories()
+
+	for _, fromID := range fromTerritories {
+		for _, toID := range toTerritories {
+			if fromID != toID && gs.AreAdjacent(fromID, toID) {

```

Comment:
from and to can be indirectly connected through other territories occupied by the player